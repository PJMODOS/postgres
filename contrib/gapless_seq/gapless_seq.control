# Gapless sequence extension
comment = 'Gapless Sequence AM'
default_version = '1.0'
module_pathname = '$libdir/gapless_seq'
relocatable = false
schema = gapless_seq
