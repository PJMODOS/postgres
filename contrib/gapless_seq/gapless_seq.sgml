<!-- doc/src/sgml/gapless_seq.sgml -->
<sect1 id="gapless_seq" xreflabel="gapless_seq">
 <title>gapless sequence</title>

 <indexterm zone="gapless_seq">
  <primary>gapless_seq</primary>
 </indexterm>

 <para>
  <filename>gapless_seq</> provides a sequence implementation that never
  has gaps in the sequence of values it produces, even after rollback or
  a crash.
 </para>

 <para>
  The consequence of this capability is that every nextval() request
  writes a WAL record recording the latest state of the sequence.
  This could be very costly and is not recommended for general
  usage except in specific applications that require this feature.
 </para>

</sect1>
