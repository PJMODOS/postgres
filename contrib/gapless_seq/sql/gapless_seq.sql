CREATE EXTENSION gapless_seq;

CREATE SEQUENCE test_gapless USING gapless;

SELECT nextval('test_gapless'::regclass);

BEGIN;
	SELECT nextval('test_gapless'::regclass);
	SELECT nextval('test_gapless'::regclass);
	SELECT nextval('test_gapless'::regclass);
ROLLBACK;

SELECT nextval('test_gapless'::regclass);

CREATE SEQUENCE test_alter_seq USING local;

SELECT nextval('test_alter_seq'::regclass);

ALTER SEQUENCE test_alter_seq USING gapless;

SELECT nextval('test_alter_seq'::regclass);

BEGIN;
	SELECT nextval('test_alter_seq'::regclass);
ROLLBACK;

SELECT nextval('test_alter_seq'::regclass);

BEGIN;
	SELECT nextval('test_alter_seq'::regclass);
	SAVEPOINT mysp;
	SELECT nextval('test_alter_seq'::regclass);
	ROLLBACK TO SAVEPOINT mysp;
	SAVEPOINT mysp2;
	SELECT nextval('test_alter_seq'::regclass);
	RELEASE SAVEPOINT mysp2;
	SELECT nextval('test_alter_seq'::regclass);
COMMIT;

ALTER SEQUENCE test_alter_seq RESTART 100 USING local;

SELECT nextval('test_alter_seq'::regclass);

-- check dump/restore
SELECT pg_sequence_get_state('test_gapless');
SELECT pg_sequence_set_state('test_gapless', pg_sequence_get_state('test_gapless'));
SELECT pg_sequence_get_state('test_gapless');

-- check that event trigger works correctly
SELECT last_value FROM gapless_seq.seqam_gapless_values ORDER BY seqid;
DROP SEQUENCE test_gapless;
SELECT last_value FROM gapless_seq.seqam_gapless_values ORDER BY seqid;

CREATE SEQUENCE test_gapless USING gapless;

-- should fail due to deps
DROP ACCESS METHOD FOR SEQUENCES gapless;
-- likewise
DROP EXTENSION gapless_seq;
-- success
DROP EXTENSION gapless_seq CASCADE;
