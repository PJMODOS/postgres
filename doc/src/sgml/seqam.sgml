<!-- doc/src/sgml/seqam.sgml -->

<chapter id="seqam">
 <title>Sequence Access Method Interface Definition</title>

  <para>
   This chapter describes the interface for writing sequence access methods.
  </para>

  <para>
   The sequence access method defines behavior of a sequence which is using it.
   Mainly how the new values are generated. Each sequence can have different
   access method in order to behave in a way that makes most sense in the
   application context.
  </para>

 <sect1 id="seqam-catalog">
  <title>Catalog Entries for Sequence Access Method</title>

  <para>
   Each sequence access method is described by a row in the
   <structname>pg_seqam</structname> system catalog (see
   <xref linkend="catalog-pg-seqam">).  The contents of a
   <structname>pg_seqam</structname> row is the name of the access method
   and are references to
   <link linkend="catalog-pg-proc"><structname>pg_proc</structname></link>
   entries that identify the functions provided by the access method.
   functions supplied by the access method.
  </para>
 </sect1>

 <sect1 id="seqam-functions">
  <title>Sequence Access Method Functions</title>

  <para>
   The behaviour of a sequence access method is defined by a set of functions.
   These functions are implemented by the sequence access method.
  </para>

  <para>
   The functions defining a sequence access method are:
  </para>

  <para>
<programlisting>
HeapTuple
seqam_init (Relation seqrel, HeapTuple tuple, int64 restart_value,
            bool restart_requested, bool is_init);
</programlisting>
   Initialize the sequence's internal state. This function is called both
   when a new sequence is created, when the <command>ALTER SEQUENCE</> command
   is called for an existing sequence or by the
   <command>TRUNCATE RESTART IDENTITY</> command.
  </para>

  <para>
   The <literal>seqrel</> points to relcache record representing the
   <link linkend="catalog-pg-class"><structname>pg_class</structname></link>
   tuple tuple of the sequence.
   <literal>tuple</> is in-memory version of the
   <structname>pg_sequence</structname> tuple itself. The
   <literal>restart_value</> is value the sequence should start from and the
   <literal>restart_requested</> gives indication if the
   <literal>restart_value</> should indeed be used. Finally <literal>is_init</>
   is <literal>true</> when this is new sequence being created or when the
   access method of a sequence has been changed, otherwise it's
   <literal>false</>.
   to the options in the <command>CREATE SEQUENCE</> or
   <command>ALTER SEQUENCE</> statement.
   <literal>reloptions</> contains parsed reloptions passed to the
   <command>CREATE SEQUENCE</> or <command>ALTER SEQUENCE</> statements.
   The <literal>values</> and <literal>nulls</> describe new tuple for the
   <structname>pg_sequence</structname> tuple which this function can update
   as needed.
  </para>

  <para>
   This function should return the updated
   <structname>pg_sequence</structname>. If the tuple does not have to be
   updated returning the input <literal>tuple</> is acceptable.
  </para>

  <para>
<programlisting>
bytea *
seqam_reloptions (ArrayType *reloptions, bool validate);
</programlisting>
   Parse and validate the reloptions array for an sequence.
   <parameter>reloptions</> is a <type>text</> array containing entries of the
   form <replaceable>name</><literal>=</><replaceable>value</>.
   The function should construct a <type>bytea</> value, which will be then sent
   to the <function>seqam_init</> and stored in the catalog.
   When <parameter>validate</> is true, the function should report a suitable
   error message if any of the options are unrecognized or have invalid
   values; when <parameter>validate</> is false, invalid entries should be
   silently ignored.  (<parameter>validate</> is false when loading options
   already stored in <structname>pg_catalog</>; an invalid entry could only
   be found if the access method has changed its rules for options, and in
   that case ignoring obsolete entries is appropriate.)
   It is OK to return NULL if default behavior is wanted.
  </para>

  <para>
<programlisting>
int64
seqam_alloc (Relation seqrel, SequenceHandle *seqh, int64 nrequested,
             int64 *last);
</programlisting>
   Allocate new sequence value(s). The <literal>nrequested</> specifies how
   many new values should be allocated by this call. Return value is the next
   allocated value and <literal>last</> should be set to last allocated value.
  </para>

  <para>
<programlisting>
bool
seqam_setval (Relation seqrel, SequenceHandle *seqh, int64 new_value)
</programlisting>
   Set the current sequence value the the <literal>new_value</>.
  </para>

  <para>
<programlisting>
ArrayType *
seqam_get_state (Relation seqrel, SequenceHandle *seqh);
</programlisting>
   Dump the current state of the sequence. The return value is one dimensional
   <literal>TEXT</> array containing list of key/value pairs it the form
   <literal>{'key1', 'value1', 'key2', 'value2'}</>. This interface is mainly
   used by <command>pg_dump</command>.
  </para>

  <para>
<programlisting>
void
seqam_set_state (Relation seqrel, SequenceHandle *seqh, ArrayType *statearr);
</programlisting>
   Restore state of the sequence based on the key/value pairs defined in
   the <literal>statearr</> in <literal>{'key1', 'value1', 'key2', 'value2'}</>
   form. The <literal>statearr</> is defined as <literal>TEXT[]</> in SQL. This
   function must accept output of the <function>seqam_get_state()</> function.
   This interface is mainly used by <command>pg_dump</command>.
  </para>

 </sect1>

 <sect1 id="seqam-storage">
  <title>Sequence Access Method Storage API</title>

  <para>
   To store the current state of the sequence, the backend provides the
   following functions which the sequence access method can use:
  </para>

  <para>
<programlisting>
void
sequence_open(Oid relid, SequenceHandle *seqh);
</programlisting>
   Open sequence with given <literal>relid</>. The <literal>seqh</> is
   output parameter which will be set to the sequence handle.
  </para>

  <para>
<programlisting>
void
sequence_close (SequenceHandle *seqh);
</programlisting>
   Release and close the opened sequence.
  </para>

  <para>
<programlisting>
HeapTuple
sequence_read_tuple (SequenceHandle *seqh);
</programlisting>
   Reads and locks the sequence tuple for processing by the access method.
   The tuple should be released by calling <function>sequence_release_tuple</>.
  </para>

<programlisting>
HeapTuple
sequence_swap_tuple (SequenceHandle *seqh, HeapTuple *newtup);
</programlisting>
   Changes the working tuple to the <literal>newtup</>. This does not change
   the sequence itself, only the state of the <struct>SequenceHandle</>. To
   save the tuple the <function>sequence_*_update()</> sequence has to be
   called. The tuple should be released by calling
   <function>sequence_release_tuple</>. Note that you can't call this function
   without calling <function>sequence_read_tuple()</> first.
  </para>

  <para>
<programlisting>
void
sequence_start_update(SequenceHandle *seqh, bool do_wal);
</programlisting>
   Start the sequence update. The <literal>do_wal</> gives hint if at least
   one of subsequent <function>sequence_apply_update()</> calls will be with
   <literal>do_wal = true</>.
</para>

  <para>
<programlisting>
void
sequence_apply_update(SequenceHandle *seqh, bool do_wal);
</programlisting>
   Save the modified sequence tuple indicating if the change should be WAL
   logged as well as saved to the catalog.
  </para>

  <para>
<programlisting>
void
sequence_finish_update(SequenceHandle *seqh, HeapTuple newtuple);
</programlisting>
   Finish the sequence update.
  </para>

  <para>
  To update the sequence tuple you have to call all three above functions in
  order to be sure the update is applied in error safe manner. The
  <function>sequence_apply_update()</> can be called multiple times between
  single pair of <function>sequence_start_update()</> and
  <function>sequence_finish_update()</> calls. Update function cannot be called
  if the tuple was already released by the
  <function>sequence_release_tuple()</> function.
  </para>

  <para>
<programlisting>
void
sequence_release_tuple(SequenceHandle *seqh);
</programlisting>
   Release the tuple read and locked by <function>sequence_read_tuple</> and/or
   added by <function>sequence_swap_tuple</>.
  </para>

 </sect1>

 <sect1 id="seqam-utility">
  <title>Sequence Access Method Utility Functions</title>

  <para>
   Additional utility functions which can be useful when writing sequence
   access methods:
  </para>

  <para>
<programlisting>
bool
sequence_needs_wal(SequenceHandle *seqh);
</programlisting>
   Returns <literal>true</> if the sequence tuple was last saved before last
   checkpoint. This can be help when deciding what to set <literal>do_wal</>
   to when calling <function>sequence_apply_update</>.
  </para>

  <para>
<programlisting>
int64
sequence_increment(Relation seqrel, int64 *value, int64 incnum,
                   int64 minv, int64 maxv, int64 incby,
                   bool is_cycled, bool report_errors);
</programlisting>
   Helper function to increment value of a sequence, correctly handling
   sequence options like min value, max value, increment value and cycling of
   the sequence value. The <literal>value</> is pointer to current value which
   should be incremented, <literal>incnum</> specifies how much should the
   value be incremented, the rest of the options should be set to values
   specified by the sequence's <structname>pg_sequence</structname> tuple.
   Returns by how much was the value increased.
  </para>

  <para>
<programlisting>
void
sequence_check_range (int64 value, int64 min_value,
                      int64 max_value, const char *valname);
</programlisting>
   Checks if the <literal>value</> is between <literal>min_value</>
   and <literal>max_value</> and raises standard error message if it's not.
   The <literal>valname</> is used as the name of the wrong value in the error
   message.
  </para>

 </sect1>

</chapter>
