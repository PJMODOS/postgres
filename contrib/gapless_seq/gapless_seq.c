/*-------------------------------------------------------------------------
 *
 * gapless_seq.c
 *
 * Portions Copyright (c) 1996-2013, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  contrib/gapless_seq/gapless_seq.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/genam.h"
#include "access/htup_details.h"
#include "access/seqam.h"
#include "access/transam.h"
#include "access/xact.h"
#include "catalog/indexing.h"
#include "catalog/namespace.h"
#include "catalog/pg_type.h"
#include "commands/sequence.h"
#include "funcapi.h"
#include "nodes/makefuncs.h"
#include "storage/procarray.h"
#include "storage/lmgr.h"
#include "utils/builtins.h"
#include "utils/fmgroids.h"
#include "utils/int8.h"
#include "utils/lsyscache.h"
#include "utils/memutils.h"
#include "utils/rel.h"


PG_MODULE_MAGIC;

/*------------------------------------------------------------
 *
 * Sequence Access Manager = Gapless functions
 *
 *------------------------------------------------------------
 */
extern Datum seqam_gapless_reloptions(PG_FUNCTION_ARGS);
extern Datum seqam_gapless_init(PG_FUNCTION_ARGS);
extern Datum seqam_gapless_alloc(PG_FUNCTION_ARGS);
extern Datum seqam_gapless_setval(PG_FUNCTION_ARGS);
extern Datum seqam_gapless_get_state(PG_FUNCTION_ARGS);
extern Datum seqam_gapless_set_state(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(seqam_gapless_reloptions);
PG_FUNCTION_INFO_V1(seqam_gapless_init);
PG_FUNCTION_INFO_V1(seqam_gapless_alloc);
PG_FUNCTION_INFO_V1(seqam_gapless_setval);
PG_FUNCTION_INFO_V1(seqam_gapless_get_state);
PG_FUNCTION_INFO_V1(seqam_gapless_set_state);

typedef struct GaplessAmdata
{
	uint32		xid;
} GaplessAmdata;

typedef struct GaplessValue
{
	Oid		seqid;
	bool	is_called;
	int64	last_value;
} GaplessValue;

#define GAPLESS_SEQ_NAMESPACE "gapless_seq"
#define VALUES_TABLE_NAME "seqam_gapless_values"
#define VALUES_TABLE_COLUMNS 3

static FormData_pg_sequence *wait_for_sequence(SequenceHandle *seqh,
											   TransactionId local_xid);
static Relation open_values_rel(void);
static HeapTuple get_last_value_tup(Relation rel, Oid seqid);
static void set_last_value_tup(Relation rel, Oid seqid, int64 last_value,
							   bool is_called, HeapTuple oldtuple);


/*
 * seqam_gapless_reloptions()
 *
 * Parse and verify the reloptions of a gapless sequence.
 */
Datum
seqam_gapless_reloptions(PG_FUNCTION_ARGS)
{
	Datum		reloptions = PG_GETARG_DATUM(0);
	bool		validate = PG_GETARG_BOOL(1);

	if (validate && PointerIsValid(DatumGetPointer(reloptions)))
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("gapless sequence does not accept any storage parameters")));

	PG_RETURN_NULL();
}

/*
 * seqam_gapless_init()
 *
 * Initialize gapless sequence
 *
 */
Datum
seqam_gapless_init(PG_FUNCTION_ARGS)
{
	Relation	seqrel = (Relation) PG_GETARG_POINTER(0);
	HeapTuple	tuple = (HeapTuple) PG_GETARG_POINTER(1);
	int64		restart_value = (int64) PG_GETARG_INT64(2);
	bool		restart_requested = (bool) PG_GETARG_BOOL(3);
	bool		is_init = (bool) PG_GETARG_BOOL(4);
	Oid			seqrelid = seqrel->rd_id;
	TupleDesc	tupDesc = RelationGetDescr(seqrel);
	Datum		amdata;
	bool		isnull;
	Relation	valrel;
	HeapTuple	valtuple;
	TransactionId		local_xid = GetTopTransactionId();
	GaplessAmdata	   *gapless_seq;

	/* Load current value if this is existing sequence. */
	valrel = open_values_rel();
	valtuple = get_last_value_tup(valrel, seqrelid);

	/*
	 * If this is new sequence or restart was provided or if there is
	 * no previous stored value for the sequence we should store it in
	 * the values table.
	 */
	if (is_init || restart_requested || !HeapTupleIsValid(valtuple))
		set_last_value_tup(valrel, seqrelid, restart_value, false, valtuple);

	/* Now we are done with values relation, but keep the lock. */
	heap_close(valrel, NoLock);

	amdata = fastgetattr(tuple, SEQ_COL_AMDATA, tupDesc,
						 &isnull);

	/* Make sure amdata is filled. */
	if (isnull)
	{
		Datum			values[SEQ_COL_LASTCOL];
		bool			nulls[SEQ_COL_LASTCOL];
		bool			replace[SEQ_COL_LASTCOL];
		struct varlena *vl = palloc0(VARHDRSZ + sizeof(GaplessAmdata));
		HeapTuple		newtup;

		memset(replace, false, sizeof(replace));
		memset(nulls, false, sizeof(nulls));
		memset(values, 0, sizeof(values));

		replace[SEQ_COL_AMDATA - 1] = true;
		nulls[SEQ_COL_AMDATA - 1] = false;

		SET_VARSIZE(vl, VARHDRSZ + sizeof(GaplessAmdata));
		values[SEQ_COL_AMDATA - 1] = PointerGetDatum(vl);

		newtup = heap_modify_tuple(tuple, tupDesc, values, nulls, replace);

		/* Don't leak memory. */
		heap_freetuple(tuple);
		tuple = newtup;

		amdata = fastgetattr(tuple, SEQ_COL_AMDATA, tupDesc,
							 &isnull);
	}

	gapless_seq = (GaplessAmdata *)
		VARDATA_ANY(DatumGetByteaP(amdata));

	/* Update the xid info */
	gapless_seq->xid = UInt32GetDatum(local_xid);

	PG_RETURN_POINTER(tuple);
}

/*
 * seqam_gapless_alloc()
 *
 * Allocate new value for gapless sequence.
 */
Datum
seqam_gapless_alloc(PG_FUNCTION_ARGS)
{
	Relation	seqrel = (Relation) PG_GETARG_POINTER(0);
	SequenceHandle *seqh = (SequenceHandle*) PG_GETARG_POINTER(1);
	/* we ignore nreguested as gapless sequence can't do caching */
	int64	   *last = (int64 *) PG_GETARG_POINTER(3);
	int64		result;
	Oid			seqrelid = RelationGetRelid(seqrel);
	Relation	valrel;
	HeapTuple	tuple;
	FormData_pg_sequence   *seq;
	GaplessAmdata	   *gapless_seq;
	TransactionId local_xid = GetTopTransactionId();

	/* Wait until the sequence is locked by us. */
	seq = wait_for_sequence(seqh, local_xid);
	gapless_seq = (GaplessAmdata *) VARDATA_ANY(&seq->amdata);

	/* Read the last value from our transactional table (if any). */
	valrel = open_values_rel();
	tuple = get_last_value_tup(valrel, seqrelid);

	/* Last value found get next value. */
	if (HeapTupleIsValid(tuple))
	{
		GaplessValue *v = (GaplessValue *) GETSTRUCT(tuple);
		result = v->last_value;

		if (v->is_called)
			(void) sequence_increment(seqrel, &result, 1, seq->min_value,
									  seq->max_value,
									  seq->increment_by,
									  seq->is_cycled, true);
	}
	else /* No last value, start from beginning. */
		result = seq->start_value;

	/*
	 * Insert or update the last value tuple.
	 */
	set_last_value_tup(valrel, seqrelid, result, true, tuple);

	/* Now we are done with values relation, but keep the lock. */
	heap_close(valrel, NoLock);

	/*
	 * If current tx is different fron the last one,
	 * update the sequence tuple as well.
	 *
	 * We don't need to WAL log the update as the only thing we save to
	 * sequence tuple is the active transaction id and we know that in case of
	 * crash the transaction id will not be active so it's ok to lose the
	 * update.
	 */
	if (gapless_seq->xid != local_xid)
	{
		sequence_start_update(seqh, true);
		gapless_seq->xid = local_xid;
		sequence_apply_update(seqh, true);
		sequence_finish_update(seqh);
	}

	*last = result;
	PG_RETURN_INT64(result);
}

/*
 * seqam_gapless_setval()
 *
 * Setval support (we don't allow setval on gapless)
 */
Datum
seqam_gapless_setval(PG_FUNCTION_ARGS)
{
	ereport(ERROR,
			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
			 errmsg("setval() is not supported for gapless sequences")));

	PG_RETURN_VOID();
}

/*
 * seqam_gapless_get_state()
 *
 * Dump state of a gapless sequence (for pg_dump)
 */
Datum
seqam_gapless_get_state(PG_FUNCTION_ARGS)
{
	Relation		seqrel = (Relation) PG_GETARG_POINTER(0);
	SequenceHandle *seqh = (SequenceHandle *) PG_GETARG_POINTER(1);
	Datum			datums[4];
	Datum			val;
	Oid				seqrelid = RelationGetRelid(seqrel);
	Relation		valrel;
	HeapTuple		tuple;
	int64			last_value;
	bool			is_called;

	/*
	 * Get the last value from the values table, if not found use start_value
	 * from the sequence definition.
	 */
	valrel = open_values_rel();
	tuple = get_last_value_tup(valrel, seqrelid);
	heap_close(valrel, RowExclusiveLock);

	if (HeapTupleIsValid(tuple))
	{
		GaplessValue *v = (GaplessValue *) GETSTRUCT(tuple);
		last_value = v->last_value;
		is_called = v->is_called;
	}
	else
	{
		FormData_pg_sequence *seq = (FormData_pg_sequence *)
			GETSTRUCT(sequence_read_tuple(seqh));
		last_value = seq->start_value;
		is_called = false;
		sequence_release_tuple(seqh);
	}

	datums[0] = CStringGetTextDatum("last_value");
	val = DirectFunctionCall1(int8out, Int64GetDatum(last_value));
	datums[1] = CStringGetTextDatum(DatumGetCString(val));

	datums[2] = CStringGetTextDatum("is_called");
	val = DirectFunctionCall1(boolout, BoolGetDatum(is_called));
	datums[3] = CStringGetTextDatum(DatumGetCString(val));

	PG_RETURN_ARRAYTYPE_P(construct_array(datums, 4, TEXTOID, -1, false, 'i'));
}

/*
 * seqam_gapless_set_state()
 *
 * Restore previously dumpred state of gapless sequence
 */
Datum
seqam_gapless_set_state(PG_FUNCTION_ARGS)
{
	Relation		seqrel = (Relation) PG_GETARG_POINTER(0);
	SequenceHandle *seqh = (SequenceHandle*) PG_GETARG_POINTER(1);
	ArrayType	   *array = PG_GETARG_ARRAYTYPE_P(2);
	Oid				seqrelid = RelationGetRelid(seqrel);
	Relation		valrel;
	HeapTuple		tuple;
	int64			last_value = 0;
	bool			is_called = false;
	bool			last_value_found = false,
					is_called_found = false;
	Datum		   *datums;
	int				count;
	int				i;
	FormData_pg_sequence   *seq;
	GaplessAmdata	   *gapless_seq;
	TransactionId local_xid = GetTopTransactionId();

	Assert(ARR_ELEMTYPE(array) == TEXTOID && ARR_NDIM(array) == 1 &&
		   (ARR_DIMS(array)[0]) % 2 == 0 && !ARR_HASNULL(array));

	deconstruct_array(array,
					  TEXTOID, -1, false, 'i',
					  &datums, NULL, &count);

	count /= 2;
	for (i = 0; i < count; ++i)
	{
		char   *key,
			   *val;

		key = TextDatumGetCString(datums[i * 2]);
		val = TextDatumGetCString(datums[i * 2 + 1]);

		if (pg_strcasecmp(key, "last_value") == 0)
		{
			last_value = DatumGetInt64(DirectFunctionCall1(int8in,
														CStringGetDatum(val)));
			last_value_found = true;
		}
		else if (pg_strcasecmp(key, "is_called") == 0)
		{
			is_called = DatumGetBool(DirectFunctionCall1(boolin,
														CStringGetDatum(val)));
			is_called_found = true;
		}
		else
			ereport(ERROR,
					(errcode(ERRCODE_SYNTAX_ERROR),
					 errmsg("invalid state key \"%s\" for gapless sequence",
							key)));
	}

	if (!last_value_found)
		ereport(ERROR,
				(errcode(ERRCODE_SYNTAX_ERROR),
				 errmsg("last_value is required parameter for gapless sequence")));

	if (!is_called_found)
		ereport(ERROR,
				(errcode(ERRCODE_SYNTAX_ERROR),
				 errmsg("is_called is required parameter for gapless sequence")));

	/* Wait until the sequence is locked by us. */
	seq = wait_for_sequence(seqh, local_xid);
	gapless_seq = (GaplessAmdata *) VARDATA_ANY(&seq->amdata);

	sequence_check_range(last_value, seq->min_value, seq->max_value, "last_value");

	/* Read the last value from our transactional table (if any). */
	valrel = open_values_rel();
	tuple = get_last_value_tup(valrel, seqrelid);

	/* Insert or update the last value tuple. */
	set_last_value_tup(valrel, seqrelid, last_value, is_called, tuple);

	/* Now we are done with values relation, but keep the lock. */
	heap_close(valrel, NoLock);

	/* Save to updated sequence. */
	sequence_start_update(seqh, true);
	gapless_seq->xid = local_xid;
	sequence_apply_update(seqh, true);
	sequence_finish_update(seqh);
	sequence_release_tuple(seqh);

	PG_FREE_IF_COPY(array, 2);

	PG_RETURN_VOID();
}

/*
 * Lock the sequence for current transaction.
 */
static FormData_pg_sequence *
wait_for_sequence(SequenceHandle *seqh, TransactionId local_xid)
{
	FormData_pg_sequence   *seq = (FormData_pg_sequence *) GETSTRUCT(sequence_read_tuple(seqh));
	GaplessAmdata	   *gapless_seq = (GaplessAmdata *) VARDATA_ANY(&seq->amdata);

	/*
	 * Read and lock the sequence for our transaction, there can't be any
	 * concurrent transactions accessing the sequence at the same time.
	 */
	while (gapless_seq->xid != local_xid &&
		   TransactionIdIsInProgress(gapless_seq->xid))
	{
		/*
		 * Release tuple to avoid dead locks and wait for the concurrent tx
		 * to finish.
		 */
		sequence_release_tuple(seqh);
		XactLockTableWait(gapless_seq->xid, NULL, NULL, XLTW_None);
		/* Reread the sequence. */
		seq = (FormData_pg_sequence *) GETSTRUCT(sequence_read_tuple(seqh));
	}

	return seq;
}

/*
 * Open the relation used for storing last value in RowExclusive lock mode.
 */
static Relation
open_values_rel(void)
{
	RangeVar   *rv;
	Oid			valrelid;
	Relation	valrel;

	rv = makeRangeVar(GAPLESS_SEQ_NAMESPACE, VALUES_TABLE_NAME, -1);
	valrelid = RangeVarGetRelid(rv, RowExclusiveLock, false);
	valrel = heap_open(valrelid, RowExclusiveLock);

	return valrel;
}

/*
 * Read the last value tuple from the values table.
 *
 * Can return NULL if tuple is not found.
 */
static HeapTuple
get_last_value_tup(Relation rel, Oid seqid)
{
	ScanKey		key;
	SysScanDesc	scan;
	HeapTuple	tuple;

	key = (ScanKey) palloc(sizeof(ScanKeyData) * 1);

	ScanKeyInit(&key[0],
				1,
				BTEqualStrategyNumber, F_OIDEQ,
				ObjectIdGetDatum(seqid)
		);

	/* FIXME: should use index */
	scan = systable_beginscan(rel, 0, true, NULL, 1, key);
	tuple = systable_getnext(scan);
	if (HeapTupleIsValid(tuple))
		tuple = heap_copytuple(tuple);
	systable_endscan(scan);

	return tuple;
}

/*
 * Insert or update the last value tuple.
 *
 * The write access to the table must be serialized by the wait_for_sequence
 * so that we don't have to have any retry scheme here.
*/
static void
set_last_value_tup(Relation rel, Oid seqid, int64 last_value, bool is_called, HeapTuple oldtuple)
{
	bool		nulls[VALUES_TABLE_COLUMNS];
	Datum		values[VALUES_TABLE_COLUMNS];
	TupleDesc	tupDesc = RelationGetDescr(rel);
	HeapTuple	tuple;

	if (!HeapTupleIsValid(oldtuple))
	{
		memset(nulls, false, VALUES_TABLE_COLUMNS * sizeof(bool));
		values[0] = ObjectIdGetDatum(seqid);
		values[1] = BoolGetDatum(is_called);
		values[2] = Int64GetDatum(last_value);

		tuple = heap_form_tuple(tupDesc, values, nulls);
		simple_heap_insert(rel, tuple);
	}
	else
	{
		bool replaces[VALUES_TABLE_COLUMNS];

		replaces[0] = false;
		replaces[1] = true;
		replaces[2] = true;

		nulls[1] = false;
		nulls[2] = false;
		values[1] = BoolGetDatum(is_called);
		values[2] = Int64GetDatum(last_value);

		tuple = heap_modify_tuple(oldtuple, tupDesc, values, nulls, replaces);
		simple_heap_update(rel, &tuple->t_self, tuple);
	}

	CatalogUpdateIndexes(rel, tuple);
}
