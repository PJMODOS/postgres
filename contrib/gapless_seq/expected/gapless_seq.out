CREATE EXTENSION gapless_seq;
CREATE SEQUENCE test_gapless USING gapless;
SELECT nextval('test_gapless'::regclass);
 nextval 
---------
       1
(1 row)

BEGIN;
	SELECT nextval('test_gapless'::regclass);
 nextval 
---------
       2
(1 row)

	SELECT nextval('test_gapless'::regclass);
 nextval 
---------
       3
(1 row)

	SELECT nextval('test_gapless'::regclass);
 nextval 
---------
       4
(1 row)

ROLLBACK;
SELECT nextval('test_gapless'::regclass);
 nextval 
---------
       2
(1 row)

CREATE SEQUENCE test_alter_seq USING local;
SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       1
(1 row)

ALTER SEQUENCE test_alter_seq USING gapless;
SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       2
(1 row)

BEGIN;
	SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       3
(1 row)

ROLLBACK;
SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       3
(1 row)

BEGIN;
	SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       4
(1 row)

	SAVEPOINT mysp;
	SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       5
(1 row)

	ROLLBACK TO SAVEPOINT mysp;
	SAVEPOINT mysp2;
	SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       5
(1 row)

	RELEASE SAVEPOINT mysp2;
	SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
       6
(1 row)

COMMIT;
ALTER SEQUENCE test_alter_seq RESTART 100 USING local;
SELECT nextval('test_alter_seq'::regclass);
 nextval 
---------
     100
(1 row)

-- check dump/restore
SELECT pg_sequence_get_state('test_gapless');
   pg_sequence_get_state    
----------------------------
 {last_value,2,is_called,t}
(1 row)

SELECT pg_sequence_set_state('test_gapless', pg_sequence_get_state('test_gapless'));
 pg_sequence_set_state 
-----------------------
 
(1 row)

SELECT pg_sequence_get_state('test_gapless');
   pg_sequence_get_state    
----------------------------
 {last_value,2,is_called,t}
(1 row)

-- check that event trigger works correctly
SELECT last_value FROM gapless_seq.seqam_gapless_values ORDER BY seqid;
 last_value 
------------
          2
          6
(2 rows)

DROP SEQUENCE test_gapless;
SELECT last_value FROM gapless_seq.seqam_gapless_values ORDER BY seqid;
 last_value 
------------
(0 rows)

CREATE SEQUENCE test_gapless USING gapless;
-- should fail due to deps
DROP ACCESS METHOD FOR SEQUENCES gapless;
ERROR:  cannot drop sequence access method gapless because extension gapless_seq requires it
HINT:  You can drop extension gapless_seq instead.
-- likewise
DROP EXTENSION gapless_seq;
ERROR:  cannot drop extension gapless_seq because other objects depend on it
DETAIL:  sequence test_gapless depends on sequence access method gapless
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
-- success
DROP EXTENSION gapless_seq CASCADE;
NOTICE:  drop cascades to sequence test_gapless
